src/puppetlabs/puppetdb/http/query.clj

```diff
@@ -304,82 +305,83 @@
+  (when log-queries?
+    (log/info (trs "Parsing PDBQuery:{0}:{1}" query-uuid (pr-str query))))
+  (let [trimmed-query (when (string? query) (str/trim query))]
+    (when (and trimmed-query (not= "" trimmed-query) (not= "null" trimmed-query))
+      (if (= (get trimmed-query 0) \[)
+        (parse-json-query trimmed-query)
+        (time-and-parse-pql query-uuid trimmed-query)))))
 
 (defn get-req->query
   "Converts parameters of a GET request to a pdb query map"
-  [{:keys [params]}
-   parse-fn]
-  (-> params
-      (update-when ["query"] parse-fn)
+  [{:keys [params globals] query-uuid :puppetdb-query-uuid :as _req}]
+  (let
+    [log-queries? (:log-queries globals)]
+    (-> params
+      (update-when ["query"] (partial parse-json-or-pql-to-ast log-queries? query-uuid))
       (update-when ["order_by"] parse-order-by-json)
       (update-when ["counts_filter"] json/parse-strict-string true)
       (update-when ["pretty"] coerce-to-boolean)
       (update-when ["include_facts_expiration"] coerce-to-boolean)
       (update-when ["include_package_inventory"] coerce-to-boolean)
-      keywordize-keys))
+      keywordize-keys)))
 
 (defn post-req->query
   "Takes a POST body and parses the JSON to create a pdb query map"
-  [req parse-fn]
-  (-> (let [req-body (request/body-string req)]
-        (try
-          (json/parse-string req-body true)
-        (catch JsonParseException e
-          (throw (IllegalArgumentException. (pprint-json-parse-exception e req-body))))))
-      (update :query (fn [query]
-                       (if (vector? query)
-                         query
-                         (parse-fn query))))))
+  [{query-uuid :puppetdb-query-uuid :as req}]
+  (let [log-queries? (get-in req [:globals :log-queries])
+        req-body (request/body-string req)]
+    (->
+     (try (json/parse-string req-body true)
+       (catch JsonParseException e
+         (throw (IllegalArgumentException. (pprint-json-parse-exception e req-body)))))
+     (update :query (fn [query]
+                      (if (vector? query)
+                        query
+                        (parse-json-or-pql-to-ast log-queries? query-uuid query)))))))
 
 (pls/defn-validated create-query-map :- puppetdb-query-schema
   "Takes a ring request map and extracts the puppetdb query from the
   request. GET and POST are accepted, all other requests throw an
   exception"
-  [req param-spec parse-fn]
+  [req param-spec]
   (convert-query-params
    (case (:request-method req)
-     :get (get-req->query req parse-fn)
-     :post (post-req->query req parse-fn)
+     :get (get-req->query req)
+     :post (post-req->query req)
      (throw (IllegalArgumentException. (tru "PuppetDB queries must be made via GET/POST"))))
    param-spec))
 
 (defn extract-query
   "Query handler that converts the incoming request (GET or POST)
   parameters/body to a pdb query map"
   ([handler param-spec]
-   (extract-query handler param-spec parse-json-query))
```

Comment:
I think we may want to keep this `parse-json-query` default because of the REST-style routes like `http://localhost:8080/pdb/query/v4/inventory`. Their only valid query content is JSON, such as
```
curl -X GET 'http://localhost:8080/pdb/query/v4/inventory' \
  --data-urlencode 'limit=1000' \
  --data-urlencode 'query=["extract", ["facts.foo"]]'
```

If we attempt to put PQL there it fails to parse as JSON
```
curl -X GET 'http://localhost:8080/pdb/query/v4/inventory' \
  --data-urlencode 'limit=1000' \
  --data-urlencode 'query=inventory[]{}'
```

<!-- name: special_char_dotted_group_by -->
```
Json parse error at line 1, column 10:

inventory[]{}
```

By removing the two parse functions, it fails instead with the error `Input to paged-sql does not match schema: ...` because it has parsed the PQL and created an invalid query
```
["from","inventory",["from","inventory",["extract",[],["=","node_state","active"]]]]
```

Reply:
You're totally right. I didn't realize that PQL queries were not allowed on entity endpoints but looking back at the handlers that makes sense. I totally got rid of the `parse-fn` removal commit and simplified the existing changes. I also added a test for that exact example ^^